### PR TITLE
Adds capabilities support for the app manager

### PR DESCRIPTION
**Pull request by bit-pirate**
New features:
- The app manager retrieves information about available capabilities from the capability server.
- Rapps with unmet capability dependencies are pruned from the list of runnable rapps.
- The app manager communicates with the capability server to start and stop capabilities as needed.
- The app manager applies remappings required for the capabilities automatically.

For the app manager to work rocon_msgs' PR 48 (https://github.com/robotics-in-concert/rocon_msgs/pull/48) needs to be merged as well.

All changes are (should be) back-wards compatible, i.e. current and new rapps are not forced to use capabilities.

### CODE REVIEW

File: rocon_app_manager/src/rocon_app_manager/caps_list.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
+#!/usr/bin/env python
```
Comment (stonier) [inline rocon_app_manager/src/rocon_app_manager/caps_list.py]: Don't need this in python modules - only useful for bash scripts.

File: rocon_app_manager/src/rocon_app_manager/caps_list.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+#
+# License: BSD
+#   https://raw.github.com/robotics-in-concert/rocon_app_platform/master/rocon_app_manager/LICENSE
```
Comment (stonier) [inline rocon_app_manager/src/rocon_app_manager/caps_list.py]: Bit messy, but we're cleaning up slowly everything to use the licence branch licenses, here you want to use:

https://raw.github.com/robotics-in-concert/rocon_app_platform/license/LICENSE

File: rocon_app_manager/src/rocon_app_manager/rapp.py [hunk 3]
```diff
@@ -299,6 +341,7 @@ def start(self, application_namespace, gateway_name, platform_info, remappings=[
                         for N in self._launch.config.nodes:
                             N.remap_args.append((t, remapped_name))
                         self._connections[connection_type].append(remapped_name)
+                        print "remapped '" + remap_to_list[indices[0]] + "' to '" + remapped_name + "'"
```
Comment (stonier) [inline rocon_app_manager/src/rocon_app_manager/rapp.py]: Probably want to drop, or comment this debugging line.

File: rocon_app_manager/src/rocon_app_manager/rapp_manager.py [hunk 4]
```diff
@@ -159,20 +171,59 @@ def _get_pre_installed_app_list(self):
...
+                rospy.logwarn("App Manager : %s" % str(e))
+                return
+            for app in self.app_list_file.available_apps:
                 if platform_compatible(platform_tuple(self.platform_info.os, self.platform_info.version, self.platform_info.system, self.platform_info.platform), app.data['platform']):
                     self.apps['pre_installed'][app.data['name']] = app
                 else:
                     rospy.logwarn('App : ' + str(app.data['name']) + ' is incompatible. App : (' + str(app.data['platform']) + ')  App Manager : (' +
                                   str(self.platform_info.os) + '.' + str(self.platform_info.version) + '.' + str(self.platform_info.system) + '.' + str(self.platform_info.platform) + ')')
 
+    def _determine_runnable_apps(self):
```
Comment (stonier) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py:191]: Do we want to include the 'platform_compatible' check in this function as well?

**Comment by stonier** [comment]
Comments in the code, the only thing that stands out from a quick look is whether the platform checking and capabilities checking can happen at the same time.

The rest of it probably needs a good test. Do you want to hold off on accepting this till it's been through a test or two?

File: rocon_app_manager/src/rocon_app_manager/rapp_manager.py (hunk 4 above)
Comment (bit-pirate) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py:191]: That was actually my first attempt. However, I don't see many similarities between both functions - one is focusing on platform,  the other on capability details. Also, I prefer to keep capability-related things together, i.e. caps compatibility check being part of the CapsList class.

File: rocon_app_manager/src/rocon_app_manager/caps_list.py (hunk 1 above)
Comment (bit-pirate) [inline rocon_app_manager/src/rocon_app_manager/caps_list.py]: Removed.

File: rocon_app_manager/src/rocon_app_manager/caps_list.py (hunk 2 above)
Comment (bit-pirate) [inline rocon_app_manager/src/rocon_app_manager/caps_list.py]: Changed.

**Comment by bit-pirate** [comment]
Integration tests for this have been added. Please merge.

File: rocon_app_manager/src/rocon_app_manager/rapp_manager.py (hunk 4 above)
Comment (stonier) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py:191]: The details are different of course and I didn't mean embedding it in the caps list class - just putting both pruning mechanisms under `_determine_runnable_apps`. 

Their details are different but they do the thing that your function says - determine runnable apps. An app isn't runnable if the platform isn't compatible. Not very logical having them separate, nor was it for me having platform check pruning pre-installed app lists, though I can't see a use case where you'd store a list of non-runnbale pre-installed apps. 

All around awkward for future devs analysing this code.

Opened a ticket for this: https://github.com/robotics-in-concert/rocon_app_platform/issues/111

Comment (stonier) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py:191]: Is there a use case for not pruning these from the pre-installed app list?

Comment (bit-pirate) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py:191]: > The details are different of course and I didn't mean embedding it in the caps list class - just putting both pruning mechanisms under _determine_runnable_apps.

Ah, misunderstood that. Bringing both under one hood seems like a good idea. I'll follow up on the new ticket.

> Is there a use case for not pruning these from the pre-installed app list?

Mainly for debugging, i.e. I created/installed new apps, but why are they not listed. Also, apps might become "runnable", if their capabilities become available. The latter should be taken into account for #111.

Comment (stonier) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py:191]: Good points.

These lists of apps are one of those things that made sense in the old app manager with our deb repositories, but got lost in the rebuild and never really made it back in yet. They need to be updated at one point if we look at app repos properly.

Comment (bit-pirate) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py:191]: > These lists of apps are one of those things that made sense in the old app manager with our deb repositories, but got lost in the rebuild and never really made it back in yet. They need to be updated at one point if we look at app repos properly.

Add a new issue for TODO and merge this one?

Comment (stonier) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py:191]: Yep, make the issue for me. I'll merge.

Comment (bit-pirate) [inline rocon_app_manager/src/rocon_app_manager/rapp_manager.py:191]: Done. New issue: https://github.com/robotics-in-concert/rocon_app_platform/issues/112

**Pull request merged**